mlir/lib/Quantum/Transforms/GridsynthPatterns.cpp

```diff
@@ -0,0 +1,524 @@
+
+    mlir::OpBuilder::InsertionGuard loopGuard(rewriter);
+    rewriter.setInsertionPointToStart(forOp.getBody());
+    mlir::Value iv = forOp.getInductionVar();
+    mlir::Value currentLoopReg = forOp.getRegionIterArg(0);
+
+    mlir::Value currentGateIndex =
+        rewriter.create<mlir::memref::LoadOp>(loc, gatesMemref, ValueRange{iv});
+
+    const int64_t numCases = pprBasis ? 19 : 10;
+    SmallVector<int64_t> caseValues;
+    caseValues.reserve(numCases);
+    for (int64_t i = 0; i < numCases; ++i) {
+        caseValues.push_back(i);
+    }
+    // Create the switch operation inside the loop
+    mlir::DenseI64ArrayAttr caseValuesAttr = rewriter.getDenseI64ArrayAttr(caseValues);
+    auto switchOp = rewriter.create<mlir::scf::IndexSwitchOp>(
+        loc, mlir::TypeRange{qregType}, currentGateIndex, caseValuesAttr, caseValues.size());
+
+    // Populate Switch Cases
+    if (pprBasis) {
+        populatePPRBasisSwitchCases(rewriter, loc, switchOp, currentLoopReg, qbitIndex);
+    }
+    else {
+        populateCliffordTSwitchCases(rewriter, loc, switchOp, currentLoopReg, qbitIndex);
+    }
+
+    // Yield the result of the switch op from the for loop
+    rewriter.setInsertionPointAfter(switchOp);
+    rewriter.create<mlir::scf::YieldOp>(loc, switchOp.getResults());
+
+    rewriter.setInsertionPointAfter(forOp);
+    mlir::Value finalReg = forOp.getResult(0);
+
+    // Return the final register and the computed runtime phase
+    rewriter.create<mlir::func::ReturnOp>(loc, mlir::ValueRange{finalReg, runtimePhase});
+
+    return func;
+}
+
+// --- Rewrite Pattern ---
+
+struct DecomposeCustomOpPattern : public mlir::OpRewritePattern<CustomOp> {
+    using mlir::OpRewritePattern<CustomOp>::OpRewritePattern;
+
+    const double epsilon;
+    const bool pprBasis;
+
+    DecomposeCustomOpPattern(mlir::MLIRContext *context, double epsilon, bool pprBasis)
+        : mlir::OpRewritePattern<CustomOp>(context), epsilon(epsilon), pprBasis(pprBasis)
+    {
+    }
+
+    mlir::LogicalResult matchAndRewrite(CustomOp op, mlir::PatternRewriter &rewriter) const override
+    {
+        StringRef gateName = op.getGateName();
+        bool isRZ = gateName == "RZ";
+        bool isPhaseShift = gateName == "PhaseShift";
+        if (!isRZ && !isPhaseShift) {
+            return failure();
+        }
+
+        mlir::Value qbitOperand;
+        SmallVector<mlir::Value> paramOperands;
+
+        for (mlir::Value operand : op->getOperands()) {
+            if (mlir::isa<catalyst::quantum::QubitType>(operand.getType())) {
+                qbitOperand = operand;
+            }
+            else {
+                paramOperands.push_back(operand);
+            }
+        }
+        if (paramOperands.size() != 1) {
+            return failure();
+        }
+        mlir::Value angle = paramOperands[0];
+        if (!qbitOperand) {
+            return failure();
+        }
```

Comment:
```suggestion
      assert(op.getQubitOperands().size() == 1 && op.getAllParams().size() == 1 && "only RZ and PhaseShift are allowed in Gridsynth decomposition");
      mlir::Value qbitOperand = op.getQubitOperands()[0];
      mlir::Value angle = op.getAllParams()[0];
```
we have convenience interface methods